codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -66,412 +70,1049 @@
+                if (childChannel.fireChannelReadPending) {
+                    // Clear early in case fireChildReadComplete() causes it to need to be re-processed
+                    childChannel.fireChannelReadPending = false;
+                    if (childChannel.fireChildReadComplete()) {
+                        flushNeeded = true;
                     }
-                });
-            } catch (Throwable cause) {
-                promise.setFailure(cause);
+                }
+                childChannel.next = null;
+                current = current.next;
             }
+        } finally {
+            // We always flush as this is what Http2ConnectionHandler does for now.
+            // TODO: I think this is not really necessary and we should be able to optimize this in the future.
+            flush0(ctx);
         }
     }
 
-    private void writeFromStreamChannel0(Object msg, boolean flush, ChannelPromise promise) {
-        try {
-            write(ctx, msg, promise);
-        } catch (Throwable cause) {
-            promise.tryFailure(cause);
-        }
-        if (flush) {
-            flush(ctx);
-        }
+    // Allow to override for testing
+    void flush0(ChannelHandlerContext ctx) {
+        flush(ctx);
     }
 
-    /**
-     * Notifies any child streams of the read completion.
-     */
-    @Override
-    public void channelReadComplete(ChannelHandlerContext ctx) {
-        for (int i = 0; i < channelsToFireChildReadComplete.size(); i++) {
-            Http2StreamChannel childChannel = channelsToFireChildReadComplete.get(i);
-            // Clear early in case fireChildReadComplete() causes it to need to be re-processed
-            childChannel.inStreamsToFireChildReadComplete = false;
-            childChannel.fireChildReadComplete();
-        }
-        channelsToFireChildReadComplete.clear();
+    // Allow to override for testing
+    void onBytesConsumed(ChannelHandlerContext ctx, Http2FrameStream stream, int bytes) throws Http2Exception {
+        consumeBytes(stream.id(), bytes);
+    }
+
+    // Allow to extend for testing
+    static class Http2MultiplexCodecStream extends DefaultHttp2FrameStream {
+
+        DefaultHttp2StreamChannel channel;
     }
 
-    ChannelFuture createStreamChannel(Channel parentChannel, EventLoopGroup group, ChannelHandler handler,
-                                              Map<ChannelOption<?>, Object> options,
-                                              Map<AttributeKey<?>, Object> attrs,
-                                              int streamId) {
-        final Http2StreamChannel channel = new Http2StreamChannel(parentChannel);
-        if (isStreamIdValid(streamId)) {
-            assert !isOutboundStream(server, streamId);
-            assert ctx.channel().eventLoop().inEventLoop();
-            channel.streamId(streamId);
+    // TODO: Handle writability changes due writing from outside the eventloop.
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)